File: Java-Core/src/main/java/presto/grammar/MatchingPatternConstraint.java
package presto.grammar;

import java.util.regex.Pattern;

import presto.error.InvalidDataError;
import presto.error.PrestoError;
import presto.expression.IExpression;
import presto.runtime.Context;
import presto.utils.CodeWriter;
import presto.value.IValue;


public class MatchingPatternConstraint implements IAttributeConstraint {

	IExpression expression;
	Pattern pattern;
	
	public MatchingPatternConstraint(IExpression expression) {
		this.expression = expression;
	}
	
	@Override
	public void checkValue(Context context, IValue value) throws PrestoError {
		if(pattern==null) {
			Object toMatch = expression.interpret(context);
			pattern = Pattern.compile(toMatch.toString());
		}
		if(!pattern.matcher(value.toString()).matches())
			throw new InvalidDataError(value.toString() + " does not match:" + pattern.pattern());
	}

	@Override
	public void toDialect(CodeWriter writer) {
		writer.append(" matching ");
		expression.toDialect(writer);
	}
}

File: Java-CodeServer/src/main/resources/js/prompto/grammar/MatchingPatternConstraint.js
var InvalidDataError = require("../error/InvalidDataError").InvalidDataError;

function MatchingPatternConstraint(expression) {
	this.expression = expression;
	this.pattern = null;
	return this;
}

MatchingPatternConstraint.prototype.checkValue = function(context, value) {
	if(this.pattern==null) {
		var toMatch = this.expression.interpret(context);
		this.pattern = new RegExp(toMatch);
	}
	if(!this.pattern.test(value.toString())) {
		throw new InvalidDataError(value.toString() + " does not match:" + this.pattern.toString());
	}
};

MatchingPatternConstraint.prototype.toDialect = function(writer) {
    writer.append(" matching ");
    this.expression.toDialect(writer);
}


exports.MatchingPatternConstraint = MatchingPatternConstraint;
